enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,386 @@
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ImagePolicySpec `json:"spec"`
+	// status contains the observed state of the resource.
+	// +optional
+	Status ImagePolicyStatus `json:"status,omitempty"`
+}
+
+// ImagePolicySpec is the specification of the ImagePolicy CRD.
+type ImagePolicySpec struct {
+	// scopes defines the list of image identities assigned to a policy. Each item refers to an image or repository in a registry implementing the "Docker Registry HTTP API V2".
+	// "scopes" uses one of the following:
+	// - complete image name, either using a tag or digest
+	// - prefixes of individual-image scopes
+	// - a wildcarded expression for matching all subdomains, the wildcard only presents at the beginning, *.example.com is a valid case, but example*.*.com is not.
+	// For more information about the format, see the document
+	// about the docker transport field:
+	// https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md#docker
+	// +kubebuilder:validation:Required
+	// +required
+	// +listType=set
+	Scopes []ImageScope `json:"scopes"`
+	// policy defines the verification policy for the items in the scopes list
+	// +kubebuilder:validation:Required
+	// +required
+	Policy Policy `json:"policy"`
+}
+
+// ImageScope is the item of the scopes list.
+// +kubebuilder:validation:Pattern=`^\*(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+$|^((((?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+(?::[0-9]+)?)|(localhost(?::[0-9]+)?))(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?::([\w][\w.-]{0,127}))?(?:@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}))?$`
+type ImageScope string
+
+// Policy defines the verification policy for the items in the scopes list.
+type Policy struct {
+	// rootOfTrust specifies the public key, the Fulcio certificate and the Rekor public key.
+	// Requires only one of keyData or fulcioCAData must be set.
+	// +required
+	RootOfTrust PolicyRootOfTrust `json:"rootOfTrust"`
+
+	// fulcioSubject specifies OIDC issuer and the email of the fulcio authentication configuration.
+	// Required if rootOfTrust.fulcioCAData is set.
+	// +optional
+	FulcioSubject PolicyFulcioSubject `json:"fulcioSubject,omitempty"`
+
+	// signedIdentity specifies what image identity the signature claims about the image.
+	// +optional
+	SignedIdentity PolicyIdentity `json:"signedIdentity,omitempty"`
+}
+
+// PolicyRootOfTrust defines the public key, the Fulcio certificate and the Rekor public key.
+type PolicyRootOfTrust struct {
+	// keyData contains inline base64 data of the public key.
+	// keyData can be empty if the image got signed keyless.
+	// Requires only one of keyData and fulcioCAData must be set.
```

Comment:
Does at least one of `keyData` and `fulcioCAData` have to be always specified? Or, can `RekorKeyData` be the only field that is set?

Reply:
Exactly one of keyData and fulcioCAData must be set. Setting only RekorKeyData is not valid.

Reply:
Can `RekorKeyData` be set for both keyData and fulcioCAData? Or, `RekorKeyData` has no use for `keyData`? If it has no use, can it have any use in the future?

Reply:
`rekorKeyData` is optional to `keyData`. The valid configuration for `rootOfTrust` is one of:
- keyData + rekorKeyData
- keyData
- fulcioCAData + rekorKeyData

Reply:
FWIW medium-term evolution here:

It might be the case that there will be a `timestampAuthorityKeyData` as an _alternative_ to `rekorKeyData`. In that case, we would accept the 3 combinations @QiWang19 lists above, + 3 more with `s/rekor/timestampAuthority/g`.

(Using a RFC 3161 timestamp server instead of Rekor is not currently implemented in c/image. But it makes more sense to me than Rekor, Cosign supports it, and I’d like to eventually add that as well.)

Reply:
What is the difference between keyData and fulcioCAData to the end user? Im wondering if they need to be separate fields, or whether we could have a keyType field that allowed the consumer to understand the difference. It would be required to be set by the user to determine which type to use, and then validation could be keyed off of this

Looking at the layout of the fields around this, I expect a discrminated union would be best that has a fulcio struct with all of the fulcio related fields present

Reply:
If we have to use the union, I guess we will have the `PolicyType` field, and this enum type allows all the valid combinations of the signature. And if more certificates are supported, we will extend the valid value of the `PolicyType` and the structs for that `PolicyType`. something like:

```go
// +union
type PolicySource struct {
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Enum=PublicKeyOnly;PublicKeyWithRekor;FulcioCA
	PolicyType         string             `json:"policyType,omitempty"`
	PublicKeyOnly      PublicKeyOnly      `json:"publicKeyOnly,omitempty"`
	PublicKeyWithRekor PublicKeyWithRekor `json:"publicKeyWithRekor,omitempty"`
	FulcioCA           FulcioCA           `json:"fulcioCA,omitempty"`
}

type PublicKeyOnly struct {
	KeyData PGPKey `json:"keyData,omitempty"`
}

type PublicKeyWithRekor struct {
	KeyData      PGPKey       `json:"keyData,omitempty"`
	RekorKeyData RekorKeyData `json:"rekorKeyData,omitempty"`
}

type FulcioCA struct {
	FulcioCAData  FulcioCAData              `json:"fulcioCAData,omitempty"`
	FulcioSubject PolicyFulcioSubject     `json:"fulcioSubject,omitempty"`
	RekorKeyData  RekorKeyData           `json:"rekorKeyData,omitempty"`
}
```

Reply:
@mtrmac @JoelSpeed  will the above struct look better?

Reply:
> What is the difference between keyData and fulcioCAData to the end user? Im wondering if they need to be separate fields, or whether we could have a keyType field that allowed the consumer to understand the difference.

I can’t see any benefit to this. The two use different formats (a public key vs. a certificate), and are used in different ways.

As is, we have two validation rules:
- ≥ 1 of `keyData` and `fulcioCAData` must be present
- ≤ 1 of `keyData` and `fulcioCAData` must be present
(or maybe that’s one “exactly 1 of…” rule, depending on how you are counting)

and other rules (`rekorKeyData` must be set if Fulcio is set) are conditioned on “is `fulcioCAData present`?”
 
With a key type, we would have:
- `keyType must be set`
- `abstractKeyData` must be set
and other rules  (`rekorKeyData` must be set if Fulcio is set) are conditioned on “is `keyType == keyTypeFulcio`?”.

So that’s about the same.

---

Using a discriminated union to separate the public-key root of trust, and the Fulcio-based root of trust, sure, why not.

Reply:
Note to self: Compare earlier discussion in https://github.com/openshift/enhancements/pull/1402#discussion_r1223488325 .

Reply:
> @mtrmac @JoelSpeed will the above struct look better?

- WRT using a discriminated union for the public key vs. Fulcio, sure, why not, I’ll defer to k8s API experts.
- Do consider that we might add a support for a RFC 3161 timestamp authority instead of Rekor (https://github.com/openshift/enhancements/pull/1402#discussion_r1330659095 ). Would Fulcio+rfc3161 be a new enum case, or a new field in the existing Fulcio case?
- On the topic of explicit discriminated unions, do we _need_ to set that up in advance for `FulcioSubject`? https://github.com/openshift/enhancements/pull/1402#discussion_r1242453618 Can that be done safely later?
- I’d prefer having a single “public key” case, with an optional Rekor field. I don’t feel strongly about this (and see above about RFC 3161 timestamps).
- I’m not too happy with the `PolicySource` name. “Source” of what? https://github.com/openshift/enhancements/pull/1402#discussion_r1242453618 names that `…SignatureAcceptanceRule`, which is too long and unergonomic.

Reply:
Names of fields can be sorted out based on SME and API reviewer conversation, that doesn't matter too much for now, but yes, a disrciminated union is the _correct_ way to structure this to be K8s like.

I would expect it to be something like (ignore names for now, we can work this out)
```
verificationPolicy:
  type: PublicKey | Fulcio
  publicKey: # Required iff type == PublicKey, forbidden otherwise
    keyData: <blah> # Required
    rekorKeyData: <blah> # optional
    ... # Whatever options are relevant here
  fulcio: # Required iff type == Fulcio, forbidden otherwise
    subject: {} # Required
    caData: <blah> # Required
    rekorKeyData: <blah> # Optional
    ... # Whatever options are relevant here
```
Lets discuss the permutation tomorrow

Reply:
updated the API to use the discriminated union